### PR TITLE
Allow custom repo override of existing source entry

### PR DESCRIPTION
**Pull request by tfoote**
From @meyerj 's [comment](https://github.com/ros-infrastructure/ros_buildfarm/issues/97#issuecomment-155903982) 

The custom repo logic is not quite right.  if the source entry is defined it will refuse to override it. 

The new implementation will print to the screen if a url is being overridden instead of aborting. If the user multiply defines the overrides it will print each successive override. 

Without the patch:
```
tfoote@yeti:/tmp/prerelease Last: [1] (2s Seconds)
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ --custom-repo hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam.git:catkin hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam2.git:catkin hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam3.git:catkin
Fetching buildfarm configuration...
Fetching rosdistro cache...
The repository 'hector_slam' appears multiple times
tfoote@yeti:/tmp/prerelease Last: [1] (1s Seconds)
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ --custom-repo hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam.git:catkin
Fetching buildfarm configuration...
Fetching rosdistro cache...
The repository 'hector_slam' appears multiple times
tfoote@yeti:/tmp/prerelease Last: [1] (2s Seconds)
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ 
Fetching buildfarm configuration...
Fetching rosdistro cache...
The repository 'hector_slam' does not have a source entry in the distribution file. We cannot generate a prerelease without a source entry.
tfoote@yeti:/tmp/prerelease Last: [1] (2s Seconds)
```

With the patch
```
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ --custom-repo hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam.git:catkin hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam2.git:catkin hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam3.git:catkin
Fetching buildfarm configuration...
Fetching rosdistro cache...
custom_repos option overiding 'hector_slam' to 'git': 'https://github.com/tu-darmstadt-ros-pkg/hector_slam2.git' 'catkin'. 
custom_repos option overiding 'hector_slam' to 'git': 'https://github.com/tu-darmstadt-ros-pkg/hector_slam3.git' 'catkin'. 
Evaluating job templates...
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/log-rotator.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_github-project.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_job-priority.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_parameters-definition.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_requeue-job.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/scm.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/scm_git.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/trigger_poll.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy_check-free-disk-space.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell_docker-info.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_check-docker.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell_key-files.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/publisher_xunit.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/publisher_mailer.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/build-wrapper_build-timeout.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/build-wrapper_timestamper.xml.em

Generated prerelease script - to execute it run:
  ./prerelease.sh
tfoote@yeti:/tmp/prerelease Last: [0] (2s Seconds)
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ --custom-repo hector_slam:git:https://github.com/tu-darmstadt-ros-pkg/hector_slam.git:catkin
Fetching buildfarm configuration...
Fetching rosdistro cache...
Evaluating job templates...
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/log-rotator.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_github-project.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_job-priority.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_parameters-definition.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/property_requeue-job.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/scm.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/scm_git.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/trigger_poll.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy_check-free-disk-space.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell_docker-info.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_check-docker.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_system-groovy.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell_key-files.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/builder_shell.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/publisher_xunit.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/publisher_mailer.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/build-wrapper_build-timeout.xml.em
/usr/lib/python2.7/dist-packages/ros_buildfarm/templates/snippet/build-wrapper_timestamper.xml.em

Generated prerelease script - to execute it run:
  ./prerelease.sh
tfoote@yeti:/tmp/prerelease Last: [0] (2s Seconds)
$ ~/work/dbf/ros_buildfarm/scripts/prerelease/generate_prerelease_script.py https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml   jade default ubuntu trusty amd64   hector_slam   --level 1   --output-dir ./ 
Fetching buildfarm configuration...
Fetching rosdistro cache...
The repository 'hector_slam' does not have a source entry in the distribution file. We cannot generate a prerelease without a source entry.
```

### CODE REVIEW

File: scripts/prerelease/generate_prerelease_script.py [hunk 1]
```diff
@@ -130,10 +130,10 @@ def main(argv=sys.argv[1:]):
         repositories[repo_name] = source_repo
 
     for repo_name, repo_type, repo_url, version in args.custom_repo:
-        if repo_name in repositories:
-            print("The repository '%s' appears multiple times" % repo_name,
+        if repo_name in repositories and repositories[repo_name]:
+            print("custom_repos option overiding '%s' to '%s': '%s' '%s'. " %
```
Comment (dirk-thomas) [inline scripts/prerelease/generate_prerelease_script.py]: Spelling: overriding

Comment (dirk-thomas) [inline scripts/prerelease/generate_prerelease_script.py]: `overriding 'name' to 'git': 'url' 'version'

"Overriding to git" sounds weird to me.

File: scripts/prerelease/generate_prerelease_script.py [hunk 2]
```diff
@@ -130,10 +130,10 @@ def main(argv=sys.argv[1:]):
...
+                  (repo_name, repo_type, repo_url, version),
                   file=sys.stderr)
-            return 1
```
Comment (dirk-thomas) [inline scripts/prerelease/generate_prerelease_script.py]: Just making it not an error doesn't seem to make use of the custom_repo later?

Comment (tfoote) [inline scripts/prerelease/generate_prerelease_script.py]: If you don't break out of the loop with the error, it gets injected into the repositories list in the next 6 lines which is what's used throughout the rest of the script.

Comment (dirk-thomas) [inline scripts/prerelease/generate_prerelease_script.py]: Ok, I didn't realize that it injects the custom rules into the repositories.

File: scripts/prerelease/generate_prerelease_script.py (hunk 1 above)
Comment (tfoote) [inline scripts/prerelease/generate_prerelease_script.py]: More clear output now: 
```
custom_repos option overriding 'hector_slam' to pull via 'git' from 'https://github.com/tu-darmstadt-ros-pkg/hector_slam2.git' with version 'catkin'. 
custom_repos option overriding 'hector_slam' to pull via 'git' from 'https://github.com/tu-darmstadt-ros-pkg/hector_slam3.git' with version 'catkin'. 
```

**Comment by dirk-thomas** [comment]
lgtm

**Pull request merged**